be2-scala/src/main/scala/ch/epfl/pop/json/MessageDataProtocol.scala

```diff
@@ -38,7 +38,49 @@ object MessageDataProtocol extends DefaultJsonProtocol {
 
   // ------------------------------- DATA FORMATTERS UTILITY ------------------------------- //
 
+  implicit object VoteElectionFormat extends RootJsonFormat[VoteElection] {
+    final private val PARAM_ID: String = "id"
+    final private val PARAM_QUESTION: String = "question"
+    final private val PARAM_VOTE: String = "vote"
+    final private val PARAM_WRITE_IN: String = "write_in"
+
+    override def read(json: JsValue): VoteElection = json.asJsObject.getFields(PARAM_ID, PARAM_QUESTION) match {
+      case Seq(id@JsString(_), question@JsString(_)) =>
+        val (voteOpt, writeInOpt): (Option[List[Int]], Option[String]) = json.asJsObject.getFields(PARAM_VOTE) match {
+          case Seq(JsArray(vote)) => (Some(vote.map(_.convertTo[Int]).toList), None)
+          case _ => json.asJsObject.getFields(PARAM_WRITE_IN) match {
+            case Seq(JsString(writeIn)) => (None, Some(writeIn))
+            case _ => throw new IllegalArgumentException(
+              s"Unable to parse vote election $json to a VoteElection object: '$PARAM_VOTE' and '$PARAM_WRITE_IN' fields are missing or wrongly formatted"
+            )
+          }
+        }
```

Comment:
this is **extremely** difficult for me to understand.
Is there no better way of expressing it, maybe sequentially (rather than nesting) ?

Reply:
Yeah sure! Can be modified. Modified this and a very similar block of code. Tell me if it's better now :)